### PR TITLE
Always show next turn option

### LINKED ISSUE #4171: Next turn is not displayed until close to turn
When routing, the name / exit of the next turn is not displayed until the current location is close to (<1 km?) the turning point. This is less of a problem for neighborhood driving, but very inconvenient for highways where knowing the exit number ahead of time is useful.

### PR DESCRIPTION
**Pull request by jm355**
Implements #4171 for android, and potentially #849

I tried to set it up for iOS, but I don't have a mac so I don't know if I did it right. I can say for sure that alwaysShowNextTurnCell needs to be uncommented. 

I also haven't been able to test it, app:compileFdroidDebugAndroidTestJavaWithJavac fails when I try to Build APKs in android studio. It built when I hit the debug button, but the emulator didn't work, and the apk it made for that wouldn't install on my phone

edit: I had to export `JAVA_HOME=/opt/android-studio/jbr` and then `./gradlew assembleFdroidBeta -Parm64` worked.

### CODE REVIEW

File: iphone/Maps/UI/Storyboard/Settings.storyboard
Comment (jm355) [inline iphone/Maps/UI/Storyboard/Settings.storyboard:1]: Uncomment this before merging

**Comment by jm355** [comment]
I don't have anywhere to drive to rn. Is there a way to do a test navigation without actually going anywhere?

**Comment by biodranik** [comment]
Thanks for your time!

We try to avoid adding unnecessary settings in Organic Maps to avoid it becoming another OSMand. For this case, a better solution looks like always showing the next turn street name/next exit number, or showing it only for highways. Let's discuss it first in the issue.

File: map/routing_manager.hpp
Comment (biodranik) [inline map/routing_manager.hpp:265]: Is there an easier way to check that the map is in landscape mode using the Framework? E.g. compare the dimensions of the map screen?

CC @vng

Comment (jm355) [reply map/routing_manager.hpp:265]: There are some spots in the code that compared dimensions, and some that used the system. Either way, I'm pretty sure that data will have to be passed through to the routingsession, so the logic should just need to be changed at the top. Note too, the dimensions will still need to be checked every time the orientation changes, so it needs to be at the top of the call stack either way

**Comment by jm355** [comment]
Can I make the font for the next turn street name bigger as part of this PR?

File: android/app/src/main/java/app/organicmaps/MwmActivity.java
Comment (jm355) [inline android/app/src/main/java/app/organicmaps/MwmActivity.java:1]: If you wanted to use the screen dimensions, you could change this to `Framework.nativeSetIsLandscape(getResources().getConfiguration().screenWidthDp > getResources().getConfiguration().screenHeightDp);`

**Comment by biodranik** [comment]
> Can I make the font for the next turn street name bigger as part of this PR?

It's better to make a separate PR.

File: map/routing_manager.hpp
Comment (biodranik) [reply map/routing_manager.hpp:265]: Please check if `void Framework::OnViewportChanged` is called on landscape/portrait rotation, and if yes, then landscape flag can be passed from there.

Comment (jm355) [reply map/routing_manager.hpp:265]: Ooh nice, yeah putting it there (and removing it from the other spots) updates it correctly. I'll commit the change shortly

File: map/routing_manager.hpp
Comment (biodranik) [inline map/routing_manager.hpp:263]: Is this function needed? `m_routingManager.RoutingSession().SetIsLandscape()` should be enough, right?

File: routing/routing_session.cpp
Comment (biodranik) [inline routing/routing_session.cpp:432]:
```suggestion
  if (distanceToTurnMeters < kShowLanesMinDistInMeters || m_route->GetCurrentTimeToNearestTurnSec() < 60.0)
```

File: routing/routing_session.cpp
Comment (biodranik) [inline routing/routing_session.cpp:430]: No need to modify variable twice.

File: routing/routing_session.cpp
Comment (biodranik) [inline routing/routing_session.cpp:442]: Please follow the convention.
```suggestion
  else if (m_isLandscape) 
```

File: routing/routing_session.cpp
Comment (biodranik) [inline routing/routing_session.cpp:742]:
```suggestion

```

**Review by biodranik — COMMENTED** [review]
Are you sure that the logic is not broken? Did you test your changes?

File: routing/routing_session.cpp
Comment (biodranik) [reply routing/routing_session.cpp:442]: What happens if m_isLandscape is false?

**Comment by jm355** [comment]
> Are you sure that the logic is not broken? Did you test your changes?

Yes, I test the changes on my phone

**Comment by jm355** [comment]
Is there anything else I can do to help get this merged?

**Comment by biodranik** [comment]
Let's test it first. Have you already done some testing?

@zyphlar can you help?

**Comment by jm355** [comment]
> Let's test it first. Have you already done some testing?
> 
> @zyphlar can you help?

Yes, I've installed it on my phone and used it for mapping. When I was many miles from the next turn, in portrait (how I usually use it) I could see the text for the next turn, then I rotated it to landscape and as expected, the next turn text went away

**Comment by jm355** [comment]
By testing it, do you mean having people with different platforms build and run, or are you looking for adding test cases to a test framework that organic maps has?

**Comment by biodranik** [comment]
We may try to experiment in production. What do you think about increasing the distance to show the next turn? Will it help in most cases? It is unlikely to remember the street or exit name that will be in 50 or 100 miles, right? Maybe increasing the distance to 3-5 miles helps?

**Comment by jm355** [comment]
Personally, I prefer to have it always shown. 

One reason is it adds consistency - some roads I turn on to aren't named, especially if it's turning into a parking lot or something. When it's always shown, I know right away whether I'll be turning onto a named road or not. And when it's not always shown, it's not obvious when it will show up. 

I recently decided to try out organic maps again after having used magic earth for a while, and it was pretty jarring to not know where I was going next, and made me think maybe the data was missing from openstreetmaps or something. It felt like I was driving somewhat blind, where I was driving for a while but had to make a mental note to keep track of how far I'd gone to remember to look at the map to know where to turn next. To me, I think designing the UI such that it minimizes the amount of times the driver has to look at the map is better, especially when it has a very minimal cost which is the case for portrait screens. So if the user can get away with looking at it once right after they make a turn and then remember the next turn even if it's 100+ miles away, that's better than forcing them to look at the map once to know how far the turn is and when they can get the turn details, and then again to see the specific turn details when they are displayed.

Another reason is at least in the US, highway exits are usually numbered. The exit number is included in this next turn string, so it's nice to see the exit number which I can easily remember, or if I'm navigating for a friend, they often want me to tell them the next exit in case I get caught up in the conversation and forget to watch the map. The nice thing about knowing the exit number is that usually each exit is about a mile apart, so then every time you pass an exit you can tell how far you are from the next turn without having to look down at the phone. 

Plus, with most devices as big as they are, the top bar doesn't take up so much space as to cause an issue. That said, I could understand if we wanted to adjust the logic so that instead of being based on whether the device is landscape, it would be based on the height of the screen. Above a certain number of pixels it's always shown, under a certain number of pixels the standard behavior (only showing the bar within a certain distance) is used.

I'm sure there are other reasons I prefer it, but those are the main ones that came to mind for me

**Comment by biodranik** [comment]
@jm355 Thanks for the explanation, it makes sense.

Do you suggest displaying an empty bar in case of an unnamed street/parking drive? Or don't display the bar with the next street at all (preferred IMO)?

Let's try to do it differently then and see how users react to this change:
1. Always show the next street even in landscape.
2. As a temporary solution for landscape, let's try to reduce the height of the top bar as much as we can, while still leaving it readable/looking more or less ok, for both iOS and Android.

**Comment by jm355** [comment]
I think if there's no name, don't display an empty bar like you say. Fortunately that's the current behavior

That sounds good to me, what do you think about making the font size bigger in portrait?

**Comment by jm355** [comment]
For minimizing the size of the top bar, are you suggesting minimizing it for both landscape and portrait, or just landscape?

**Comment by jm355** [comment]
I've updated the top bar size for landscape mode on android, but I don't know how to do it for iphone

**Comment by biodranik** [comment]
@jm355 can you please provide two screenshots, before and after your changes, to quickly compare and see the change?

Only landscape mode needs improvements.

**Comment by biodranik** [comment]
@kirylkaveryn can you please help and decrease the height of the top bar with the "Next street" in the nav mode on iOS in a separate PR?

**Comment by zyphlar** [comment]
I've downloaded the WebDebug and will test on my drive today

**Comment by zyphlar** [comment]
My first bit of feedback is that we should combine this with the color/style change of the top bar (someone else made it green instead of white etc) *and* we should make the blue arrow indicator the same color (green) so that it visually links together. It's useful to have a pane that says "in 2 miles turn right onto Street" but we want to design correctly so nobody thinks this is a "current street" informational display.

**Comment by jm355** [comment]
> My first bit of feedback is that we should combine this with the color/style change of the top bar (someone else made it green instead of white etc) _and_ we should make the blue arrow indicator the same color (green) so that it visually links together. It's useful to have a pane that says "in 2 miles turn right onto Street" but we want to design correctly so nobody thinks this is a "current street" informational display.

I agree

**Comment by kirylkaveryn** [comment]
@jm355 @biodranik 
I reduced the top bar insets from 24(top from safe area)/12(bottom) to 10/10. So size was decreased by 16pt.
Are you think it's enough?

Before/ after.

Vert:
image

image


Hor:
image

image

**Comment by biodranik** [comment]
Definitely looks better. Can the top safe area be also reused somehow? E.g., on devices with the safe area, can the intent from the screen edge to the text be the same, as on devices without the safe area?

**Comment by kirylkaveryn** [comment]
> E.g., on devices with the safe area, can the intent from the screen edge to the text be the same, as on devices without the safe area?

@biodranik it's highly not recommend. Because the safe area insets may vary from zero (devices with Home button) up to about 56.
And a device may have different safe area insets from top and bottom (devices with Dynamic Island).
So it requires a lot if-elses to properly handle all screen traits.

**Comment by biodranik** [comment]
@kirylkaveryn can safe area (if present) be reused to display the text?

What are the examples of devices/screens with larger/stranger safe area insets to keep in mind when designing the UI?

**Comment by kirylkaveryn** [comment]
@biodranik yes we can. But its bad UI to use this space for important elements.
Documentation says:
> The safe area defines the area within a view that _isn’t covered by a navigation bar, tab bar, toolbar, or other views a view controller might provide.
This is safe area in our case:
image
And some examples:
image
image


image

**Comment by Jean-BaptisteC** [comment]
Nice works, navigation screen is really better in landscape

**Comment by jm355** [comment]
Is the last thing holding this back changing the background color of the top bar?

**Comment by jm355** [comment]
For landscape mode, I think the ideal is more like this, with the top/bottom bar only on one side of the screen, which would be beyond the scope of this PR (https://github.com/organicmaps/organicmaps/issues/2177)

File: map/routing_manager.hpp
Comment (Jean-BaptisteC) [inline map/routing_manager.hpp:262]: Remove?

**Comment by jm355** [comment]
Is the last thing holding this back changing the background color of the top bar, or is that beyond the scope of this pr?

**Comment by biodranik** [comment]
@rtsisyk this one should be thoroughly tested in the alpha/beta

I've noticed that other navigators _do not always show_ the next turn. And that is for a reason: when that next turn appears, a driver sees that the turn is coming soon, and prepares for it. If it's not visible yet, then it's safe to relax and drive for a longer distance.

**Comment by jm355** [comment]
google maps

**Comment by jm355** [comment]
Magic earth 
Screenshot_20240406-000927

**Comment by Jean-BaptisteC** [comment]
Yes, we have planned to refactor routing screen. But I prefer doing this in a separate PR.

**Comment by jm355** [comment]
Any updates on this, or anything I can do to help get this integrated? This is a big enough deal to me that I'm sticking with magic earth until it's changed, even though I'd rather be using organic maps otherwise

**Comment by biodranik** [comment]
Let's rebase it and fix build errors in checks.

**Comment by jm355** [comment]
Looks like all the checks pass

File: android/app/src/main/res/layout-land/layout_nav_top.xml
Comment (biodranik) [inline android/app/src/main/res/layout-land/layout_nav_top.xml:42]: Should it be hard-coded or made relative in some other way? How was this value determined? Will it work with a very large font set in system settings?

Comment (jm355) [reply android/app/src/main/res/layout-land/layout_nav_top.xml:42]: This file was just copied from https://github.com/organicmaps/organicmaps/blob/656686cbb4f768d0c28313e7921aa555992a68eb/android/app/src/main/res/layout/layout_nav_top.xml, there are only three lines that are different, and they're in the first FrameLayout section:
```
~ $ diff layout_nav_top.xml layout_land_nav_top.xml 
21c21
<       android:layout_height="@dimen/nav_street_height"
---
>       android:layout_height="wrap_content"
27c27
<         android:layout_height="match_parent"
---
>         android:layout_height="wrap_content"
30c30
<         android:layout_gravity="center_vertical"
---
>         android:layout_gravity="center_horizontal"
```
I don't remember the exact reason I changed them, I think it was based on some of the requests to make landscape look different to portrait mode. Personally I think changes to landscape might fit better in a different PR (and would be better off being done by someone who understands the android layout xml better), but that was what was requested so I figured out how to make it look different

**Review by biodranik — APPROVED** [review]
Ok, thanks, let's move on and test it live in betas and then decide if it's ok and how we can improve.

**Comment by jm355** [comment]
Awesome, thank you!!

**Comment by fabwu** [comment]
> @fabwu can you please check its behavior on iOS?

@biodranik I'm a bit lost here. What exactly should I test?

**Comment by biodranik** [comment]
What happens with the next turn street, is it displayed or not )

**Comment by zyphlar** [comment]
@fabwu in this PR, the next turn you need to take will always be displayed at the top of the screen, instead of only a few feet/meters ahead of time. So an easy way to test is to build and run this PR's code, open the app, choose to navigate to some place, and it should instantly say something like "Main Street" with an arrow and distance (if Main Street was the closest road to your position) even if you're not moving and not close to the turn. (If the distance is <500m maybe you should try navigating down a long straight road without turns to verify that the turn name is still shown always even when the next turn is far away)

**Comment by fabwu** [comment]
> @fabwu in this PR, the next turn you need to take will always be displayed at the top of the screen, instead of only a few feet/meters ahead of time. So an easy way to test is to build and run this PR's code, open the app, choose to navigate to some place, and it should instantly say something like "Main Street" with an arrow and distance (if Main Street was the closest road to your position) even if you're not moving and not close to the turn. (If the distance is <500m maybe you should try navigating down a long straight road without turns to verify that the turn name is still shown always even when the next turn is far away)

Ok thanks. I currently don't have much internet but I'll get back to you with my test results asap.

**Comment by fabwu** [comment]
@zyphlar I tested your change in the simulator and it looks good to me. I get the street/exit name right after I started navigation. Should we close the to issue referenced in this PR?

**Comment by zyphlar** [comment]
@fabwu let's make reminders for ourselves to test the next beta with this feature on Android and iOS and close that issue if it actually looks solved. Remember it should even show next turns when they're many miles away, not just close turns

**Pull request merged**
